### PR TITLE
Ignore missing roots in fingerprints

### PR DESCRIPTION
**Pull request by wolfs**
Instead of having some fingerprinting strategies which consider missing roots and some which ignore them, this PR ensures all fingerprinting strategies ignore missing roots.

This is an extended version of https://github.com/gradle/gradle/pull/19877, which only ignored missing roots for relative path fingerprinting.

### CODE REVIEW

File: subprojects/core/src/test/groovy/org/gradle/internal/fingerprint/impl/AbsolutePathFileCollectionFingerprinterTest.groovy [hunk 1]
```diff
@@ -63,21 +63,21 @@ class AbsolutePathFileCollectionFingerprinterTest extends Specification {
...
+        fingerprint.fingerprints.keySet().collect { new File(it) } == [file, dir, dir2, file2]
     }
 
     def "retains order of elements in the snapshot"() {
         given:
         TestFile file = tmpDir.createFile('file1')
-        TestFile file2 = tmpDir.file('file2')
-        TestFile file3 = tmpDir.file('file3')
+        TestFile noExist = tmpDir.file('file2')
+        TestFile file3 = tmpDir.createDir('file3')
```
Comment (lptr) [inline subprojects/core/src/test/groovy/org/gradle/internal/fingerprint/impl/AbsolutePathFileCollectionFingerprinterTest.groovy:73]: Call this `dir` if it's a directory?

File: subprojects/core/src/test/groovy/org/gradle/internal/fingerprint/impl/AbsolutePathFileCollectionFingerprinterTest.groovy [hunk 2]
```diff
@@ -63,21 +63,21 @@ class AbsolutePathFileCollectionFingerprinterTest extends Specification {
...
-        fingerprint.fingerprints.keySet().collect { new File(it) } == [file, dir, dir2, file2, noExist]
+        fingerprint.fingerprints.keySet().collect { new File(it) } == [file, dir, dir2, file2]
     }
 
     def "retains order of elements in the snapshot"() {
         given:
         TestFile file = tmpDir.createFile('file1')
-        TestFile file2 = tmpDir.file('file2')
-        TestFile file3 = tmpDir.file('file3')
+        TestFile noExist = tmpDir.file('file2')
```
Comment (lptr) [inline subprojects/core/src/test/groovy/org/gradle/internal/fingerprint/impl/AbsolutePathFileCollectionFingerprinterTest.groovy:72]: How about `missing`?

Comment (lptr) [reply subprojects/core/src/test/groovy/org/gradle/internal/fingerprint/impl/AbsolutePathFileCollectionFingerprinterTest.groovy:72]: In other tests we also keep the variable name and the file name the same where it makes sense, I think we should follow that here too.

Comment (wolfs) [reply subprojects/core/src/test/groovy/org/gradle/internal/fingerprint/impl/AbsolutePathFileCollectionFingerprinterTest.groovy]: I renamed the variable and the file names.

**Review by lptr — APPROVED** [review]
LGTM.

**Review by asodja — APPROVED** [review]

**Comment by wolfs** [comment]
See #14587.

**Pull request closed without merging**